lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,728 @@
+// Metatypes can be converted to Class when they are metatypes for concrete
+// classes. https://github.com/apple/swift/pull/27479#discussion_r344418131
+clang::QualType ClangTypeConverter::visitMetatypeType(MetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+// TODO: [stronger-checking-in-clang-type-conversion]
+// Existential metatypes where the base is a non-metatype existential can be
+// converted to Class<P, Q, ...> when the protocols are all ObjC.
+// https://github.com/apple/swift/pull/27479#discussion_r344418131
+clang::QualType
+ClangTypeConverter::visitExistentialMetatypeType(ExistentialMetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitClassType(ClassType *type) {
+  auto &clangCtx = ClangASTContext;
+  auto swiftDecl = type->getDecl();
+
+  // TODO: [non-objc-class-clang-type-conversion]
+  // See the corresponding note in GenClangType.cpp
+  if (!swiftDecl->isObjC())
+    return getClangIdType(clangCtx);
+
+  // TODO: As an optimization, we could check if swiftDecl has a clang decl
+  // already attached to it and if so, reuse it instead of making a new one.
+
+  // produce the clang type INTF * if it is imported ObjC object.
+  clang::IdentifierInfo *ForwardClassId =
+    &clangCtx.Idents.get(swiftDecl->getName().get());
+  auto *CDecl = clang::ObjCInterfaceDecl::Create(
+                        clangCtx, clangCtx.getTranslationUnitDecl(),
+                        clang::SourceLocation(), ForwardClassId,
+                        /*typeParamList*/nullptr, /*PrevDecl=*/nullptr,
+                        clang::SourceLocation());
+
+  // Attach an objc_runtime_name attribute with the Objective-C name to use
+  // for this class.
+  SmallString<64> runtimeNameBuffer;
+  CDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                   CDecl->getASTContext(),
+                   swiftDecl->getObjCRuntimeName(runtimeNameBuffer)));
+
+  auto clangType  = clangCtx.getObjCInterfaceType(CDecl);
+  return clangCtx.getObjCObjectPointerType(clangType);
+}
+
+// TODO: We should try to preserve type arguments on imported ObjC generic
+// classes, instead of relying on our knowledge of clang's encoding.
+// This would entail extracting the type arguments, calling `convert` to
+// create clang types, extracting the ObjCProtocolDecls and then using
+// getObjCObjectType with `id` as the base.
+clang::QualType
+ClangTypeConverter::visitBoundGenericClassType(BoundGenericClassType *type) {
+  // Any @objc class type in Swift that shows up in an @objc method maps 1-1 to
+  // "id <SomeProto>"; with clang's encoding ignoring the protocol list.
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitBoundGenericType(BoundGenericType *type) {
+  // The only possibilities are *Pointer<T>, SIMD*<T>,
+  // ImplicitlyUnwrappedOptional<T> and Optional<T>.
+
+  if (!type->getDecl()->isObjC())
+    return clang::QualType();
```

Comment:
Literally none of the above types are ObjC, so I'm curious how any of this works.

Reply:
Modified the wrong function. I was working on `visitBoundGenericClassType`, which is right above this one.

Reply:
As you point out, it almost certainly was not working, but we are not testing that yet. 😞 When I land the attribute, we can test all these out.